dashboard/test/models/pd/session_test.rb

```diff
@@ -100,9 +100,113 @@ class Pd::SessionTest < ActiveSupport::TestCase
     refute session_past.too_soon_for_attendance?
     assert session_past.too_late_for_attendance?
 
-    session_ended = create :pd_session, workshop: workshop_ended
-    refute session_ended.open_for_attendance?
+    session_ended = create(:pd_session, workshop: workshop_ended).tap(&:assign_code)
+    assert session_ended.open_for_attendance?
     refute session_ended.too_soon_for_attendance?
-    assert session_ended.too_late_for_attendance?
+    refute session_ended.too_late_for_attendance?
+  end
+
+  test 'attendance link for one day workshop' do
+    #current day is Saturday
+    #workshop started Friday
+    workshop_a = create :pd_workshop, started_at: Time.now - 1.day
+    #session starts on Monday
+    session_a = create(:pd_session, workshop: workshop_a, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
+    refute session_a.open_for_attendance?
+    assert session_a.show_link?
+
+    #current day is Friday
+    #workshop started Friday
+    workshop_a1 = create :pd_workshop, started_at: Time.now
+    #session starts on Monday
+    session_a1 = create :pd_session, workshop: workshop_a1, start: Time.now + 3.days
+    refute session_a1.open_for_attendance?
+    refute session_a1.show_link?
+    assert session_a1.too_soon_for_attendance?
+    refute session_a1.too_late_for_attendance?
+  end
+
+  test 'attendance link see immediately' do
+    #current day is Saturday
+    #workshop started Saturday
+    workshop_b = create :pd_workshop, started_at: Time.now
+    #session starts on Monday
+    session_b = create(:pd_session, workshop: workshop_b, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
+    refute session_b.open_for_attendance?
+    assert session_b.show_link?
+  end
+
+  test 'attendance links for three day workshop' do
+    #current day is Saturday
+    #workshop started Saturday
+    workshop_c = create :pd_workshop, started_at: Time.now
+    #session 1 starts on Monday, 2 on Tuesday, 3 on Wednesday
+    session_c1a = create(:pd_session, workshop: workshop_c, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
+    session_c2b = create(:pd_session, workshop: workshop_c, start: Time.now + 3.days - 1.minute).tap(&:assign_code)
+    session_c3c = create(:pd_session, workshop: workshop_c, start: Time.now + 4.days - 1.minute).tap(&:assign_code)
```

Comment:
These sessions don't need to be created individually. See the `num_sessions` and `sessions_from` attributes on the `pd_workshop` factory [here](https://github.com/code-dot-org/code-dot-org/blob/a4ad67204234dae0c9d13076071ba006a665bc8a/dashboard/test/factories/pd_factories.rb#L24-L26).

Reply:
That's super helpful thanks!